### ISSUE TITLE
Why does a SQLModel class with `table=True` not validate data ?

### ISSUE THREAD

**Issue by deajan** [tiangolo/sqlmodel#453]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from pydantic import BaseModel
from sqlmodel import SQLModel, Field


class BaseModelTest(BaseModel):
    id: int
    some_bool: bool
    desc: str

class SQLModelTest(SQLModel, table=True):
    id: int = Field(default=None, primary_key=True)
    some_bool: bool
    desc: str


# This will work as expected
sqlmodel_instance = SQLModelTest(id=3, some_bool=False, desc="This will work")
print(sqlmodel_instance)
# This will work as expected
pydantic_instance = BaseModelTest(id=3, some_bool=False, desc="This will work")
print(pydantic_instance)


# This will work but should not
sqlmodel_instance = SQLModelTest(id=3, some_bool="blob", desc=False)
print(sqlmodel_instance)
# This will not work as expected
pydantic_instance = BaseModelTest(id=3, some_bool="blob", desc=False)
print(pydantic_instance)
```

### Description

When creating a SQLModel class instance, I thought that I'd have all the advantages of pydantic and SQLAlchemy classes toghether.
I was quite surprised when I realized that SQLModel classes created with `table=True` do not validate data as a pydantic class would do (see my example code).
Having read about all SQLModel documentation, I didn't find anywhere this is stated.
Especially in https://sqlmodel.tiangolo.com/tutorial/fastapi/multiple-models/ it would make sense to explain that a `table=True` won't do data validation as a pydantic class would do.

Maybe I did not understand the purpose of SQLModel enough, but this is quite disturbing for me.
I think it would be really nice to update the documentation regarding this particular point.

### Operating System

Windows

### SQLModel Version

0.0.8

### Python Version

Python 3.10.7 x64

### Additional Context

PS: Sorry, I'm comming here from Tortoise-ORM, and it's my first time using SQLModel, so I cannot guarantee that I will be quite helpful resolving issues here yet ;) I'll do my best.

**Comment by phi-friday** [comment]
To guess, when used as a table object, there is no verification process because it is used to map data that has already been verified in database.
(i think because ```SQLModel``` use ```DeclarativeMeta``` when table=True. [link](https://github.com/tiangolo/sqlmodel/blob/75ce45588b6a13136916929be4c44946f7e00cbd/sqlmodel/main.py#L214))
In particular, when calling many objects from database, there is no unnecessary verification process, so there is a performance advantage.
Of course, if the verification process is essential, ```SQLModel.validate``` method can be used.

I've asked the same question before, but I'm just using ```SQLModel.validate``` method for now. I want to know why, too.

**Comment by deajan** [comment]
I do understand that too, reads from database don't need validation. But writes do. Nevertheless it's not really clear in the documentation.
And since SQLModel is supposed to reduce code duplication, having a class for ORM and a class for Pydantic validation.... doesn't help deduplicate.

**Comment by phi-friday** [comment]
@deajan
I don't know the answer to why, but I found a simple way to solve it.
 [link](https://github.com/tiangolo/sqlmodel/issues/52#issuecomment-1225746421)

**Comment by 5p4k** [comment]
I think this is answered here: https://github.com/tiangolo/sqlmodel/issues/52#issuecomment-1311987732

For future reference, I found this while looking for why `model_validate_json` did not properly validate for a `SQLModel` with `table=True`.

**Comment by mg3146** [comment]
I agree - this just totally threw me off. I expected it to be fully validating my table data.  why bother inheriting from BaseModel at all if not. 

> This feature/bug should be fully documented, as I (and many others) expect the SQL model to validate data. Only after doing a lot of digging to find out why one of my field_validator wasn't working did I come across this GitHub issue.

**Comment by iloveitaly** [comment]
This was also very surprising to me. [Looks like this PR may fix the issue](https://github.com/fastapi/sqlmodel/pull/1041), although it's unclear if it's the right approach.

**Comment by fastkits** [comment]
when update new version ?

**Comment by ml31415** [comment]
> when update new version ?

I don't think, that there will be a version fixing this. It rather looks like, that this oddity is considered a feature. Sadly.

**Comment by SDVoGo** [comment]
That's not what I expected.
I have been looking for a solution in the docs (in SQLmodel and Pydantic too) for 3 days now.
Why would I expect an error from the DB rather than my backend application?

I mean, I don't consider myself a senior so I guess whoever is working on it knows better than me but:
If the intended behavior is waiting for the DB to raise an error. Why are my tests still committed?

My problem was really stupid.

``` python
class Formula(SQLModel, table=True):
    id: int | None = Field(default=None, primary_key=True)
    name: str
    formula: str
```
`name` and `formula` are strings not None. I accidentally found that empty string was ok for both but I didn't like that.

I can't use `@model_validator(mode="after")` because won't consider any validators.
Then the only way is to spam ` [field name]: str = Field(min-lenght=1)`.
I really can't understand

**Comment by rainergo** [comment]
The whole point for me of considering SQLModel was to avoid duplication in my data model (ORM AND Pydantic type check of data model). If this is not a bug, then SQLModel is totally useless.

**Comment by Samreay** [comment]
I see I'm late to this party after many hours spent banging my head against a wall in my code base. The fascinating part is that the model validators are called with `model_validate` but not `model_validate_json`. In case its useful at all for disambiguation, here's what I had written up to post as the replication for a new issue before finding this one:

```python
from datetime import datetime as dt
from datetime import timezone as tz
from typing import Annotated, Self

from pydantic import BeforeValidator
from sqlmodel import Field, SQLModel

IntDatetime = Annotated[
    dt,
    BeforeValidator(lambda x: dt.fromtimestamp(x / 1000, tz=tz.utc) if isinstance(x, int) else x),
]


class ExampleWorks(SQLModel):
    time: IntDatetime = Field(primary_key=True)


class ExampleBroken(SQLModel, table=True):
    time: IntDatetime = Field(primary_key=True)


class NastyBase(SQLModel):
    @classmethod
    def model_validate_json(cls, json: str | bytes) -> Self:
        model = super().model_validate_json(json)
        return cls.model_validate(model)


class ExampleBrokenFixed(NastyBase, table=True):
    time: IntDatetime = Field(primary_key=True)


print("Without a table, validation works:")
print(ExampleWorks(time=dt(2025, 8, 14, tzinfo=tz.utc)))
print(ExampleWorks.model_validate({"time": 1755129600000}))
print(ExampleWorks.model_validate_json('{"time": 1755129600000}'))

print("\nMaking it a table breaks validation:")
print(ExampleBroken(time=dt(2025, 8, 14, tzinfo=tz.utc)))
print(ExampleBroken.model_validate({"time": 1755129600000}))
print(ExampleBroken.model_validate_json('{"time": 1755129600000}'))

print("\nForcing revalidation makes this slow but it technically works:")
print(ExampleBrokenFixed(time=dt(2025, 8, 14, tzinfo=tz.utc)))
print(ExampleBrokenFixed.model_validate({"time": 1755129600000}))
print(ExampleBrokenFixed.model_validate_json('{"time": 1755129600000}'))

``` 
Outputs:
```
Without a table, validation works:
time=datetime.datetime(2025, 8, 14, 0, 0, tzinfo=datetime.timezone.utc)
time=datetime.datetime(2025, 8, 14, 0, 0, tzinfo=datetime.timezone.utc)
time=datetime.datetime(2025, 8, 14, 0, 0, tzinfo=datetime.timezone.utc)

Making it a table breaks validation:
time=datetime.datetime(2025, 8, 14, 0, 0, tzinfo=datetime.timezone.utc)
time=datetime.datetime(2025, 8, 14, 0, 0, tzinfo=datetime.timezone.utc)
time=1755129600000

Forcing revalidation makes this slow but it technically works:
time=datetime.datetime(2025, 8, 14, 0, 0, tzinfo=datetime.timezone.utc)
time=datetime.datetime(2025, 8, 14, 0, 0, tzinfo=datetime.timezone.utc)
time=datetime.datetime(2025, 8, 14, 0, 0, tzinfo=datetime.timezone.utc)
```

**Comment by m-danya** [comment]
Well, it was really unpleasant to discover... The whole point of the library, to combine pydantic and sqlalchemy, seems to have failed.

This example prints `<class 'str'>`:

```python
from datetime import datetime
from fastapi import FastAPI
from sqlmodel import Field, SQLModel
import uvicorn

app = FastAPI()


class FavoriteDate(SQLModel, table=True):
    id: int | None = Field(default=None, primary_key=True)
    date: datetime


@app.post("/date")
async def add_date(favorite_date: FavoriteDate):
    print(type(favorite_date.date))


uvicorn.run(app, port=5000)

```

**Comment by spandan-sharma** [comment]
Pasting the [TL;DR from here](https://github.com/fastapi/sqlmodel/issues/52#issuecomment-3244702008) 

---

## TL;DR

You don't need multi-models. Here's a complete solution to address the problems:

```
from contextlib import contextmanager
from contextvars import ContextVar
from typing import Self

import sqlmodel as sm

_ONGOING_MODEL_VALIDATE: ContextVar[bool] = ContextVar('_ONGOING_MODEL_VALIDATE', False)

@contextmanager
def set_ongoing_model_validate():
    token = _ONGOING_MODEL_VALIDATE.set(True)
    yield
    _ONGOING_MODEL_VALIDATE.reset(token)

class OurBase(sm.SQLModel):
    model_config = {'validate_assignment': True}

    def __init__(self, **data):
        if self.model_config.get('table', False) and not _ONGOING_MODEL_VALIDATE.get():
            self_copy = self.model_copy()
            self.__pydantic_validator__.validate_python(data, self_instance=self_copy)
            data = self_copy.model_dump()
            self.__dict__ |= self_copy.__dict__
        super().__init__(**data)

    @classmethod
    def model_validate(cls: type[Self], *args, **kwargs) -> Self:
        with set_ongoing_model_validate():
            return super().model_validate(*args, **kwargs)
```

Derive from this base (modify `model_config` etc if needed) and you should be fine. Given the age, I'm not hopeful for any official solution landing, so I'm just using this and moving on. To know why/how it works, read on.

**Comment by leogout** [comment]
> Pasting the [TL;DR from here](https://github.com/fastapi/sqlmodel/issues/52#issuecomment-3244702008)
> 
> ## TL;DR
> You don't need multi-models. Here's a complete solution to address the problems:
> `[...]`
> Derive from this base (modify `model_config` etc if needed) and you should be fine. Given the age, I'm not hopeful for any official solution landing, so I'm just using this and moving on. To know why/how it works, read on.

To me, the whole point of using a library is to avoid digging in its protected attributes and private methods. Running undocumented code in an untested way based on a comment in a Github thread dosen't feel appealing. 

**This** is the first paragraph of the README : 

> It combines SQLAlchemy and Pydantic and tries to simplify the code you write as much as possible, allowing you to reduce the code duplication to a minimum, but while getting the best developer experience possible.

I surely don't understand all the pros and cons of this library as it is used by a lot of people, but my usecase, which I thought simple, isn't covered. Seeing the hundreds of reactions and the number of issues opened on this subject, I would say something isn't working as people expected. Maybe it should be made clear from the start that this usecase is not covered, and that people will need to write separate classes extending eachother to manage the validation and the orm separately. The README says it "combines" pydantics and SQLAlchemy but it is unclear to me how it combines them.

In any case, thank you for your work, provided for free to me and others. I hope I'll find another usecase for it.

**Comment by MarleTangible** [comment]
I just would like to make this absolutely clear.

The library claims ([statement](https://sqlmodel.tiangolo.com/#sqlalchemy-and-pydantic)) to combine pydantic and sqlalchemy models, but in its very basic use-case it doesn't validate the request against the model when used with FastAPI which defeats the purpose.

Even the code from the [simple api tutorial](https://sqlmodel.tiangolo.com/tutorial/fastapi/simple-hero-api/#create-heroes-path-operation) returns a 500 internal error instead of a 422 validation error. This is the very first endpoint users create with the library, so either the library description should be updated with corrected claims, or it should be made clear from get-go that defining a class with `table=True` is not validated due to these (listed) reasons.

``` Python
# great simple way of processing the Hero class
@app.post("/heroes/")
def create_hero(hero: Hero):
    with Session(engine) as session:
        session.add(hero)
        session.commit()
        session.refresh(hero)
        return hero
```

Please don't get me wrong, the library is great in the fact that I don't need to deal with sqlalchemy types and simply create a pydantic model and interact with the database. But I still need a separate model to first validate the user input. The final list of models becomes something like below, see my comments please.

``` Python
# common fields
class HeroBase(SQLModel):
    name: str = Field(index=True)
    secret_name: str
    age: int | None = Field(default=None, index=True)

# database model
class Hero(HeroBase, table=True):
    id: int | None = Field(default=None, primary_key=True)

# model for post requests
class HeroCreate(HeroBase):
    pass

# model for responses
# practically the same as Hero but reasons ...
class HeroPublic(HeroBase):
    id: int
```

**Comment by shubhamyx** [comment]
"Hi, I've opened a PR to address the documentation gap mentioned here: fastapi/sqlmodel#1880"

**Comment by Jamim** [comment]
> "Hi, I've opened a PR to address the documentation gap mentioned here: [#1880](https://github.com/fastapi/sqlmodel/pull/1880)"

Hey @shubhamyx,
Quote marks in your message tell us you are an AI agent 💯 
Why don't you disclose that in your profile?
